src/hooks/useGameplay.ts

```diff
@@ -0,0 +1,185 @@
+import React from 'react';
+import {
+    districts,
+    provinces,
+} from '#utils/constants';
+
+import {
+    Challange,
+    GameMode,
+    GameState,
+    Region,
+} from '#types';
+
+const DEFAULT_TICK_INTERVAL = 500;
+
+const regionMap: {
+    [key in GameMode]: Region[];
+} = {
+    district: districts,
+    province: provinces,
+};
+
+const getRandomRegions = (mode: GameMode, amount: number) => {
+    const regions = regionMap[mode];
+    const randomRegions = [...Array(amount).keys()].map(() => {
+        const randomIndex = Math.floor(Math.random() * Math.floor(regions.length));
+        const region = regions[randomIndex];
+
+        return {
+            code: region.code,
+            title: region.title,
+        };
+    });
+
+    return randomRegions;
+};
+
+export function useTimer(
+    use: boolean,
+    tickInterval:number = DEFAULT_TICK_INTERVAL,
+) {
+    const [tick, setTick] = React.useState(0);
+    const initialTimeRef = React.useRef(new Date().getTime());
+    const [initialTime, setInitialTime] = React.useState(initialTimeRef.current);
+    const [initialLapTime, setInitialLapTime] = React.useState(initialTimeRef.current);
+    const [elapsed, setElapsed] = React.useState(0);
+    const [lapElapsed, setLapElapsed] = React.useState(0);
+    const shouldTick = React.useRef(use);
+
+    const addLap = React.useCallback(() => {
+        setInitialLapTime(new Date().getTime());
+        setLapElapsed(0);
+    }, [setInitialLapTime, setLapElapsed]);
+
+    React.useEffect(() => {
+        shouldTick.current = use;
+        if (use) {
+            setInitialTime(new Date().getTime());
+        }
+    }, [use, shouldTick, setInitialTime]);
+
+    const handleTick = React.useCallback(() => {
+        const { current } = shouldTick;
+
+        if (current) {
+            const now = new Date().getTime();
+            const currentElapsed = now - initialTime;
+            const currentLapElapsed = now - initialLapTime;
+            setTick((prevTick) => prevTick + 1);
+            setElapsed(currentElapsed);
+            setLapElapsed(currentLapElapsed);
+        }
+    }, [initialTime, initialLapTime, setTick, setElapsed, setLapElapsed, shouldTick]);
```

Comment:
We never include useRef values in useEffect list. Because, the reference never changes